cg/meta/clean/clean_flow_cells.py

```diff
@@ -0,0 +1,179 @@
+"""An API that handles the cleaning of flow cells."""
+import logging
+import time
+from pathlib import Path
+from typing import List, Optional
+
+from housekeeper.store.models import File
+
+from cg.apps.housekeeper.hk import HousekeeperAPI
+from cg.constants import SequencingFileTag
+from cg.constants.time import TWENTY_ONE_DAYS
+from cg.exc import HousekeeperBundleVersionMissingError, HousekeeperFileMissingError
+from cg.meta.demultiplex.housekeeper_storage_functions import (
+    get_sample_sheets_from_latest_version,
+)
+from cg.models.demultiplex.flow_cell import FlowCellDirectoryData
+from cg.store import Store
+from cg.store.models import Flowcell, SampleLaneSequencingMetrics
+from cg.utils.files import remove_directory_and_contents
+from cg.utils.time import is_directory_older_than_days_old
+
+LOG = logging.getLogger(__name__)
+
+
+class CleanFlowCellAPI:
+    """
+            Handles the cleaning of flow cells in the flow cells and demultiplexed runs directories.
+    Requirements for cleaning:
+            Flow cell is older than 21 days
+            Flow cell is backed up
+            Flow cell is in StatusDB
+            Flow cell has sequencing metrics in StatusDB
+            Flow cell has fastq files in Housekeeper
+            Flow cell has SPRING files in Housekeeper
+            Flow cell has SPRING metadata in Housekeeper
+            Flow cell has a sample sheet in Housekeeper
+    """
+
+    def __init__(
+        self, status_db: Store, housekeeper_api: HousekeeperAPI, flow_cell_path: Path, dry_run: bool
+    ):
+        self.status_db: Store = status_db
+        self.hk_api: HousekeeperAPI = housekeeper_api
+        self.flow_cell = FlowCellDirectoryData(flow_cell_path=flow_cell_path)
+        self.current_time: float = time.time()
+        self.dry_run: bool = dry_run
+
+    def delete_flow_cell_directory(self) -> bool:
+        """
+        Delete the flow cell directory if it fulfills all requirements.
+        """
+        is_error_raised: bool = False
+        try:
+            self.set_sample_sheet_path_from_housekeeper()
```

Comment:
The sample sheet path is not automatically set... I added it here as it can cause a *HousekeeperBundleVersionMissing* error. Not nice to do it this way, likely the most efficient.

Reply:
I think if all the conditions for cleaning a flow cell are met, the sample sheet must be in HK. I think for now this could be here but I think if the sample sheet is not in HK an error would have been raised much earlier in the demux flow

Reply:
Yes, the issue is that the sample sheet is not set automatically in `FlowCellDirectoryData` that is why I needed to make this function here. Not sure how to best solve his. I think there was an effort to refactor this. But I definetly agree with you. This should not happen at this point.